### PR TITLE
Allow GitHub `workflow_dispatch` events to publish

### PR DESCRIPTION
**Pull request by spydon**
This PR allows new versions of packages to be published from GitHub `workflow_dispatch` events.
The same restriction on the `refType` is in place, it still needs to be run targeting a tag.

Closes: #7177

Reason for change:
This is currently a blocker for actions like the [melos-action](https://github.com/marketplace/actions/melos-action) to be used to its full potential, where the user can create releases directly from the GitHub UI in the following manner:

1. The maintainer presses  ["Run workflow"](https://docs.github.com/en/actions/using-workflows/manually-running-a-workflow).
image
2. The action creates a branch with updated `CHANGELOG.md` and `pubspec.yaml` files and uploads it as a PR
3. A maintainer reviews and merges the release PR
4. Another workflow is triggered by the merge that then tags the packages and triggers yet another workflow through `workflow_dispatch`, since tags created from a GitHub action [aren't allowed to trigger other actions](https://docs.github.com/en/actions/using-workflows/triggering-a-workflow#triggering-a-workflow-from-a-workflow)
5. The last workflow uploads the newly versioned packages to pub.dev

This is a lot less error prone than doing manual releases from the command line, especially in bigger monorepos.

---

- [x] I’ve reviewed the contributor guide and applied the relevant portions to this PR.

<details>
  <summary>Contribution guidelines:</summary><br>

- See our [contributor guide](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md) for general expectations for PRs.
- Larger or significant changes should be discussed in an issue before creating a PR.
- Contributions to our repos should follow the [Dart style guide](https://dart.dev/guides/language/effective-dart) and use `dart format`.
- Most changes should add an entry to the changelog and may need to [rev the pubspec package version](https://github.com/dart-lang/sdk/wiki/External-Package-Maintenance#making-a-change).
- Changes to packages require [corresponding tests](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md#Testing).

Note that many Dart repos have a weekly cadence for reviewing PRs - please allow for some latency before initial review feedback.
</details>

### CODE REVIEW

**Comment by jonasfj** [comment]
Are you sure these checks will still work?
```dart
    if (agent.payload.refType != 'tag') {
      throw AuthorizationException.githubActionIssue(
          'publishing is only allowed from "tag" refType, this token has "${agent.payload.refType}" refType');
    }
    final expectedRefStart = 'refs/tags/';
    if (!agent.payload.ref.startsWith(expectedRefStart)) {
      throw AuthorizationException.githubActionIssue(
          'publishing is only allowed from "refs/tags/*" ref, this token has "${agent.payload.ref}" ref');
    }
    final expectedTagValue = tagPattern.replaceFirst('{{version}}', newVersion);
    if (agent.payload.ref != 'refs/tags/$expectedTagValue') {
```

Will it require that the workflow_dispatch is triggered on a tag?

**Comment by spydon** [comment]
@jonasfj added a test in https://github.com/dart-lang/pub-dev/pull/7462/commits/1be34bf71bfd10ed3a9ea8d8440b17d319f524fa

I'm not sure if it is possible to trigger the second `if`, possibly by manipulating the request somehow, because if the `refType` is `tag` the `ref` must surely start with `refs/tags` right? Anyhow, if anyone would manage to do that it is still covered, just like it was before.

File: app/lib/package/backend.dart [hunk 1]
```diff
@@ -1268,7 +1268,8 @@ class PackageBackend {
 
     // TODO: consider allowing other events from
     //       https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows
-    if (agent.payload.eventName != 'push') {
+    if (agent.payload.eventName != 'push' &&
+        agent.payload.eventName != 'workflow_dispatch') {
       throw AuthorizationException.githubActionIssue(
           'publishing is only allowed from "push" events, this token originates from a "${agent.payload.eventName}" event');
```
Comment (clragon) [inline app/lib/package/backend.dart:1274]: We probably wanna adjust this error message

**Review by clragon — CHANGES_REQUESTED** [review]

File: app/lib/package/backend.dart (hunk 1 above)
Comment (spydon) [reply app/lib/package/backend.dart]: Good catch, updated!

**Comment by Gustl22** [comment]
@spydon seems, there are some conflicts.

**Comment by spydon** [comment]
> @spydon seems, there are some conflicts 😸 I'm excited about this feature ;D

It's just the changelog, it has to be updated every time something else is merged (gives me throwbacks from before we used Melos in all our projects :sweat_smile:).

**Comment by Gustl22** [comment]
@clragon @jonasfj anything blocking to finish the PR? Best regards :)

**Comment by spydon** [comment]
> @clragon @jonasfj anything blocking to finish the PR? Best regards :)

The design document needs to be reviewed and accepted first, can be seen here: https://github.com/dart-lang/pub-dev/issues/7177#issuecomment-1977676681

And hopefully after that there only needs to be a pair of check boxes added and theeen we can  get this in. 😄

**Comment by isoos** [comment]
@spydon: We've come to a conclusion and we'd like to allow this event, but only as a UI option after an admin allowed it on the admin UI. Would you be interested to implement it as such?

It would need to be stored on the [`GithubPublishingConfig`](https://github.com/dart-lang/pub-dev/blob/master/pkg/_pub_shared/lib/data/package_api.dart#L69) object, and be displayed on the [UI of the package page admin](https://github.com/dart-lang/pub-dev/blob/master/app/lib/frontend/templates/views/pkg/admin_page.dart#L240).

@jonasfj: I think we could use a simple `bool? isWorkflowDispatchEnabled` flag on the object and a checkbox on the UI - or would you like to see a more open text field for the additional event names?

**Comment by jonasfj** [comment]
@isoos I think we should do two booleans.

```dart
bool isFromWorkflowDispatchEventEnabled;
bool isFromPushEventEnabled;
```

Or maybe `allowPublishFrom<Event>`, but we probably need to make two checkboxes in the UI:
 * [x] Enable publishing from `push` events.
 * [ ] Enable publishing from `workflow_dispatch` events.

Where the first defaults to being enabled, and the second can be enabled.

I think the ability to disallow publishing from `push` events is nice. I think we can allow the freedom of letting people enable both, but let's give them the ability to only allow the thing they use.

**Comment by jonasfj** [comment]
@isoos this is why I thought it might be necessary with some data migration, in order to just create the properties on all entities. That's something we could easily do with a backfill job, right?

**Comment by isoos** [comment]
> @isoos this is why I thought it might be necessary with some data migration, in order to just create the properties on all entities. That's something we could easily do with a backfill job, right?

Yeah, makes sense.

**Comment by spydon** [comment]
> Would you be interested to implement it as such?

Sounds great! 😃

**Comment by isoos** [comment]
Note: I'm working on a PR that will add the checkboxes to the UI + other changes and tests.

**Comment by isoos** [comment]
Closing as #7766 was merged.

**Pull request closed without merging**